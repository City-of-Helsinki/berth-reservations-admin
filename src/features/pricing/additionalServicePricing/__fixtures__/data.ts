import { AdditionalServicePricingProps } from '../AdditionalServicePricing';
import {
  ProductServiceType,
  PriceUnits,
  PeriodType,
  AdditionalProductType,
  AdditionalProductTaxEnum,
} from '../../../../@types/__generated__/globalTypes';

export const data: AdditionalServicePricingProps['data'] = {
  __typename: 'AdditionalProductNodeConnection',
  edges: [
    {
      __typename: 'AdditionalProductNodeEdge',
      node: {
        __typename: 'AdditionalProductNode',
        id: 'a8400b59-534c-4934-b643-083a4273ca1a',
        service: ProductServiceType.DINGHY_PLACE,
        priceValue: 10,
        priceUnit: PriceUnits.AMOUNT,
        period: PeriodType.SEASON,
        productType: AdditionalProductType.OPTIONAL_SERVICE,
        taxPercentage: AdditionalProductTaxEnum.TAX_24_00,
      },
    },
  ],
};
