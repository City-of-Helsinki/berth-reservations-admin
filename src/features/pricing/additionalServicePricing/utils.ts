import { AdditionalServicePricingProps, AdditionalService } from './AdditionalServicePricing';

export const getAdditionalServiceData = (data: AdditionalServicePricingProps['data']): AdditionalService[] => {
  if (!data) return [];

  return data.edges.reduce<AdditionalService[]>((acc, edge) => {
    if (!edge?.node) return acc;

    const berthPrice = {
      id: edge.node.id,
      service: edge.node.service,
      price: edge.node.priceValue,
      tax: edge.node.taxPercentage,
      period: edge.node.period,
    };

    return [...acc, berthPrice];
  }, []);
};
