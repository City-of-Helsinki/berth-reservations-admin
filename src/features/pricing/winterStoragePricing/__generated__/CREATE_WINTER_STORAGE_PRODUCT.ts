/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { CreateWinterStorageProductMutationInput, PriceUnits, PlaceProductTaxEnum } from "./../../../../@types/__generated__/globalTypes";

// ====================================================
// GraphQL mutation operation: CREATE_WINTER_STORAGE_PRODUCT
// ====================================================

export interface CREATE_WINTER_STORAGE_PRODUCT_createWinterStorageProduct_winterStorageProduct {
  __typename: "WinterStorageProductNode";
  id: string;
  priceValue: any;
  priceUnit: PriceUnits;
  taxPercentage: PlaceProductTaxEnum;
}

export interface CREATE_WINTER_STORAGE_PRODUCT_createWinterStorageProduct {
  __typename: "CreateWinterStorageProductMutationPayload";
  winterStorageProduct: CREATE_WINTER_STORAGE_PRODUCT_createWinterStorageProduct_winterStorageProduct | null;
}

export interface CREATE_WINTER_STORAGE_PRODUCT {
  createWinterStorageProduct: CREATE_WINTER_STORAGE_PRODUCT_createWinterStorageProduct | null;
}

export interface CREATE_WINTER_STORAGE_PRODUCTVariables {
  input: CreateWinterStorageProductMutationInput;
}
