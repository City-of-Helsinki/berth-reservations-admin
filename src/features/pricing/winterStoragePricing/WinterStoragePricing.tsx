import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Row } from 'react-table';
import { useMutation } from '@apollo/react-hooks';

import Card from '../../../common/card/Card';
import Section from '../../../common/section/Section';
import CardHeader from '../../../common/cardHeader/CardHeader';
import CardBody from '../../../common/cardBody/CardBody';
import Table, { Column, COLUMN_WIDTH } from '../../../common/table/Table';
import { formatPrice } from '../../../common/utils/format';
import EditForm, { EDIT_FORM_TYPE } from '../editModal/EditForm';
import Text from '../../../common/text/Text';
import { WinterStoragePricing as WinterStoragePricingData } from './__generated__/WinterStoragePricing';
import { getWinterStorageData } from './utils';
import { PeriodType } from '../../../@types/__generated__/globalTypes';
import { getPeriodTKey } from '../../../common/utils/translations';
import EditModal from '../editModal/EditModal';
import { UPDATE_WINTER_STORAGE_PRICE_MUTATION, CREATE_WINTER_STORAGE_PRODUCT_MUTATION } from './mutations';
import {
  UPDATE_WINTER_STORAGE_PRICE,
  UPDATE_WINTER_STORAGE_PRICEVariables as UPDATE_WINTER_STORAGE_PRICE_VARS,
} from './__generated__/UPDATE_WINTER_STORAGE_PRICE';
import {
  CREATE_WINTER_STORAGE_PRODUCT,
  CREATE_WINTER_STORAGE_PRODUCTVariables as CREATE_WINTER_STORAGE_PRODUCT_VARS,
} from './__generated__/CREATE_WINTER_STORAGE_PRODUCT';

export interface WinterStoragePrice {
  id: string;
  productId: string | undefined | null;
  area: string | undefined | null;
  privateCustomer: number | undefined;
  company: number | undefined;
  period: PeriodType;
}

export interface WinterStoragePricingProps {
  data: WinterStoragePricingData | undefined | null;
  loading: boolean;
  className?: string;
}

const WinterStoragePricing = ({ data, loading, className }: WinterStoragePricingProps) => {
  const { t, i18n } = useTranslation();
  const [editRowValues, setEditRowValues] = useState<WinterStoragePrice>();
  const [updateWinterStoragePrice] = useMutation<UPDATE_WINTER_STORAGE_PRICE, UPDATE_WINTER_STORAGE_PRICE_VARS>(
    UPDATE_WINTER_STORAGE_PRICE_MUTATION
  );
  const [createWinterStoragePrice] = useMutation<CREATE_WINTER_STORAGE_PRODUCT, CREATE_WINTER_STORAGE_PRODUCT_VARS>(
    CREATE_WINTER_STORAGE_PRODUCT_MUTATION
  );

  const winterStorageCols: Column<WinterStoragePrice>[] = [
    {
      Header: t('pricing.winterStorage.area') || '',
      accessor: 'area',
    },
    {
      Header: t('pricing.winterStorage.privateCustomer') || '',
      accessor: 'privateCustomer',
      Cell: ({ cell: { value } }) => (value ? formatPrice(value, i18n.language) : '-'),
    },
    {
      Header: t('pricing.winterStorage.company') || '',
      accessor: 'company',
      Cell: ({ cell: { value } }) => (value ? formatPrice(value, i18n.language) : '-'),
    },
    {
      Header: t('pricing.winterStorage.period') || '',
      accessor: 'period',
      Cell: ({ cell }) => t(getPeriodTKey(cell.value)),
    },
    {
      id: 'edit',
      Header: t('common.edit') || '',
      sortType: 'none',
      width: COLUMN_WIDTH.S,
      Cell: ({ row }: { row: Row<WinterStoragePrice> }) => (
        <button onClick={() => setEditRowValues(row.original)}>
          <Text color="brand">{t('common.edit')}</Text>
        </button>
      ),
    },
  ];

  const handleSubmit = async ({
    productId,
    id: winterStorageAreaId,
    privateCustomer: priceValue,
  }: WinterStoragePrice) => {
    if (productId) {
      await updateWinterStoragePrice({
        variables: { input: { id: productId, priceValue } },
      });
    } else {
      await createWinterStoragePrice({
        variables: { input: { winterStorageAreaId, priceValue } },
      });
    }

    setEditRowValues(undefined);
  };

  const handleClose = () => setEditRowValues(undefined);

  return (
    <>
      <Card className={className}>
        <CardHeader title={t('pricing.winterStorage.title')} />
        <CardBody>
          <Section>{t('pricing.winterStorage.description')}</Section>
          <Table
            columns={winterStorageCols}
            data={getWinterStorageData(data)}
            loading={loading}
            theme="basic"
            renderEmptyStateRow={() => t('common.notification.noData.description')}
          />
        </CardBody>
      </Card>
      <EditModal isOpen={!!editRowValues} closeModal={handleClose}>
        {editRowValues && (
          <EditForm
            closeModal={handleClose}
            formType={EDIT_FORM_TYPE.WINTER_STORAGE}
            initialValues={editRowValues}
            onSubmit={handleSubmit}
          />
        )}
      </EditModal>
    </>
  );
};

export default WinterStoragePricing;
