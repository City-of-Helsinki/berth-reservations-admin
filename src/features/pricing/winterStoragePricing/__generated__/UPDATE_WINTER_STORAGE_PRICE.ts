/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { UpdateWinterStorageProductMutationInput, PriceUnits, PlaceProductTaxEnum } from "./../../../../@types/__generated__/globalTypes";

// ====================================================
// GraphQL mutation operation: UPDATE_WINTER_STORAGE_PRICE
// ====================================================

export interface UPDATE_WINTER_STORAGE_PRICE_updateWinterStorageProduct_winterStorageProduct {
  __typename: "WinterStorageProductNode";
  id: string;
  priceValue: any;
  priceUnit: PriceUnits;
  taxPercentage: PlaceProductTaxEnum;
}

export interface UPDATE_WINTER_STORAGE_PRICE_updateWinterStorageProduct {
  __typename: "UpdateWinterStorageProductMutationPayload";
  winterStorageProduct: UPDATE_WINTER_STORAGE_PRICE_updateWinterStorageProduct_winterStorageProduct | null;
}

export interface UPDATE_WINTER_STORAGE_PRICE {
  updateWinterStorageProduct: UPDATE_WINTER_STORAGE_PRICE_updateWinterStorageProduct | null;
}

export interface UPDATE_WINTER_STORAGE_PRICEVariables {
  input: UpdateWinterStorageProductMutationInput;
}
