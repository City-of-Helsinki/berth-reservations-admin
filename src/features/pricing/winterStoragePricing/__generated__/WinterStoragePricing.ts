/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { PriceUnits } from "./../../../../@types/__generated__/globalTypes";

// ====================================================
// GraphQL fragment: WinterStoragePricing
// ====================================================

export interface WinterStoragePricing_edges_node_properties_product {
  __typename: "WinterStorageProductNode";
  priceValue: any;
  priceUnit: PriceUnits;
}

export interface WinterStoragePricing_edges_node_properties {
  __typename: "WinterStorageAreaProperties";
  name: string | null;
  product: WinterStoragePricing_edges_node_properties_product | null;
}

export interface WinterStoragePricing_edges_node {
  __typename: "WinterStorageAreaNode";
  id: string;
  properties: WinterStoragePricing_edges_node_properties | null;
}

export interface WinterStoragePricing_edges {
  __typename: "WinterStorageAreaNodeEdge";
  node: WinterStoragePricing_edges_node | null;
}

export interface WinterStoragePricing {
  __typename: "WinterStorageAreaNodeConnection";
  edges: (WinterStoragePricing_edges | null)[];
}
