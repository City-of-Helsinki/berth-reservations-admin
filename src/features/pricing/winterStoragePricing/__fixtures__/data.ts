import { WinterStoragePricingProps } from '../WinterStoragePricing';
import { PriceUnits } from '../../../../@types/__generated__/globalTypes';

export const data: WinterStoragePricingProps['data'] = {
  __typename: 'WinterStorageAreaNodeConnection',
  edges: [
    {
      __typename: 'WinterStorageAreaNodeEdge',
      node: {
        __typename: 'WinterStorageAreaNode',
        id: 'b3cb95f7-e9ce-438a-8447-203145f047fe',
        properties: {
          __typename: 'WinterStorageAreaProperties',
          name: 'Corporate Developer Mobility',
          product: {
            __typename: 'WinterStorageProductNode',
            priceValue: '10.2',
            priceUnit: PriceUnits.AMOUNT,
          },
        },
      },
    },
  ],
};
