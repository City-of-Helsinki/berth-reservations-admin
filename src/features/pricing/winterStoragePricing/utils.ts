import { WinterStoragePricing as WinterStoragePricingData } from './__generated__/WinterStoragePricing';
import { WinterStoragePrice } from './WinterStoragePricing';
import { PeriodType } from '../../../@types/__generated__/globalTypes';

export const getWinterStorageData = (data: WinterStoragePricingData | undefined | null): WinterStoragePrice[] => {
  if (!data) return [];

  return data.edges.reduce<WinterStoragePrice[]>((acc, edge) => {
    if (!edge?.node) return acc;

    const priceValue = edge.node.properties?.product?.priceValue;
    const privateCustomer = Number.isNaN(priceValue) ? undefined : priceValue;
    const company = Number.isNaN(priceValue) ? undefined : priceValue * 2;

    const winterStorageArea = {
      id: edge.node.id,
      productId: edge.node.properties?.product?.id,
      area: edge.node.properties?.name,
      privateCustomer,
      company,
      period: PeriodType.SEASON,
    };

    return [...acc, winterStorageArea];
  }, []);
};
