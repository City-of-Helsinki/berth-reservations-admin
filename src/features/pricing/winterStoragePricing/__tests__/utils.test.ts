import { getWinterStorageData } from '../utils';
import { WinterStoragePricing } from '../__generated__/WinterStoragePricing';
import { data } from '../__fixtures__/data';

describe('utils', () => {
  describe('getWinterStorageData', () => {
    it('should return an array of WinterStoragePrice objects', () => {
      expect(getWinterStorageData(data)).toMatchSnapshot();
    });

    it("company's fee should be twice as much as the private customer's", () => {
      const berthPrices = getWinterStorageData(data as WinterStoragePricing);

      expect(berthPrices[0].company).toBe((berthPrices[0].privateCustomer as number) * 2);
    });

    it('should return an empty array when the provided data is undefined or null', () => {
      expect(getWinterStorageData(undefined)).toEqual([]);
      expect(getWinterStorageData(null)).toEqual([]);
    });

    it('should remove all potential null/undefined nodes', () => {
      const data = { __typename: '', edges: [null] };

      expect(getWinterStorageData(data as WinterStoragePricing)).toEqual([]);
    });
  });
});
