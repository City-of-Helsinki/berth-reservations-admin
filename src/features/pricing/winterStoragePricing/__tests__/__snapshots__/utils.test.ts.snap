// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getWinterStorageData should return an array of BerthPrice objects 1`] = `
Array [
  Object {
    "area": "Corporate Developer Mobility",
    "company": 20.4,
    "id": "b3cb95f7-e9ce-438a-8447-203145f047fe",
    "period": "SEASON",
    "privateCustomer": "10.2",
  },
]
`;

exports[`utils getWinterStorageData should return an array of WinterStoragePrice objects 1`] = `
Array [
  Object {
    "area": "Corporate Developer Mobility",
    "company": 20.4,
    "id": "b3cb95f7-e9ce-438a-8447-203145f047fe",
    "period": "SEASON",
    "privateCustomer": "10.2",
  },
]
`;
